Add plans toggle between Builder and Hosting plans

dzver opened the pull request:
<!--
Link a related Github/Linear issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword. Consider using "part of" instead.
-->

Part of #

## Proposed Changes

<img width="1443" height="823" alt="Screenshot 2025-08-28 at 15 48 08" src="https://github.com/user-attachments/assets/bec07013-ed35-48ea-8197-239e0938b109" />
<img width="1443" height="823" alt="Screenshot 2025-08-28 at 15 48 17" src="https://github.com/user-attachments/assets/b0a7239d-4eb8-44f3-8627-5b2f08183659" />
<img width="1443" height="823" alt="Screenshot 2025-08-28 at 15 48 25" src="https://github.com/user-attachments/assets/de383ec8-7881-41c0-bcef-b899a88e9d68" />
<img width="1443" height="823" alt="Screenshot 2025-08-28 at 15 49 16" src="https://github.com/user-attachments/assets/d7310f9e-87c7-4ab8-9ddd-5b7cf89aa09c" />


## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

*

## Testing Instructions

* Find the calypso_plans_page_visual_separation_2025_09 in Abacus (22324-explat-experiment)
* Create 3 bookmarklets with control, hosting, and builder variations
* Create a test user and assign it to one of them (you'll need to test all 3)
* Start the signup onboarding or reload the domains step, and then move on to plans

Notes:
* There are two general cases, when a free domain is selected or not
* When a domain is selected, there's no Free plan in the builder tab, and there's a link above, which is not bold
* When a domain is not selected, there is a Free plan in the builder tab but there is no link above the hosting tab (this might be a design flaw, we may need to reconsider)

Control should remain unchanged

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

dzver committed:
Add a basic toggle on top of Plans

dzver committed:
A few more configs

dzver committed:
Show the deemphasize option and hide the Free plan

It's unstyled but there

dzver committed:
Styling

dzver committed:
Move the toggle above the stepper header

dzver committed:
Match Figma

dzver committed:
Align the copy to match Figma

dzver committed:
Hide pills in visual split

dzver committed:
Update the default intent

dzver committed:
Add right margin on mobile

dzver committed:
Update header and subheader according to the interim copy

dzver committed:
Add the new hosting features to the plans grid

dzver committed:
Add titles

dzver committed:
Update three features based on figma

dzver committed:
Sync the hosting features with figma

dzver committed:
Update the commerce features

Still unordered but reached my time limit

dzver committed:
Show free per request by Isa

dzver committed:
hide the second subheader

dzver committed:
Add ... or start with free to the new copy

dzver committed:
Fix the Free plan button to be inline

dzver committed:
Adjust pixels and colors to the latest changes

dzver committed:
Fine-tune borders, spacings

dzver committed:
Remove features per latest design request

dzver committed:
Undo changes that leaked to control

dzver committed:
Add experiment code

dzver committed:
Introduce the third variation

dzver committed:
Show toggle

dzver committed:
Term selection toggle showed up after a rebase

dzver committed:
Update features per latest request

Local dev environmnent no longer in use
Bring back 2 features

dzver committed:
Add tracks event

dzver committed:
Toggle simplification

dzver committed:
Nuria's preferred loading is with no selection

dzver committed:
Adjust styles per Nuria's request

claudiucelfilip commented on client/landing/stepper/declarative-flow/internals/steps-repository/unified-plans/unified-plans-step.tsx:
```diff
@@ -551,7 +595,20 @@ function UnifiedPlansStep( {
-					heading={ <Step.Heading text={ getHeaderText() } subText={ fallbackSubHeaderText } /> }
+					heading={
+						<>
+							{ ( visualSplitVariation === 'default_websitebuilder' ||
+								visualSplitVariation === 'default_hosting' ) && (
```
Nit pick: I see we're checking intent for the variation - would make to check for it here as well?

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/20026059

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @dzver for including a screenshot in the description! This is really helpful for our translators.

dzver committed:
Fix a bug that Claudiu found in the plans grid

Clicking the intent tab on smaller screens messed up the comparison grid

Claudiu also proposed the solution

chriskmnds commented on client/landing/stepper/declarative-flow/internals/steps-repository/unified-plans/unified-plans-step-styles.scss:
```diff
@@ -66,3 +68,14 @@
 	position: relative;
 }
+
+.step-container-v2__visual-split {
```
I believe the `step-container-v2` suite of classes is meant to be for the Stepper wireframes. I'm not sure if it's a good idea to be prefixing classes with that name, as it reduces clarity e.g. "visual-split" is not part of the wireframe.

chriskmnds commented on client/landing/stepper/declarative-flow/internals/steps-repository/unified-plans/unified-plans-step-styles.scss:
```diff
@@ -66,3 +68,14 @@
+	margin-top: -17px;
+
+	// Target the last TR with class plans-grid-next-features-grid__feature-group-row
+	tr.plans-grid-next-features-grid__feature-group-row:last-child {
```
Ideally, we shouldn't be referencing anything from an external package like plans-grid-next in the step styles. It is also unclear from here what it tries to do. This could best be handled via a property or condition inside the plans grid, instead IMO.

chriskmnds commented on client/lib/domains/use-visual-split-experiment.ts:
It's unclear why this is defined under `domains` folder, since it's used for the plans step/code? 🤔

chriskmnds commented on client/my-sites/plans-features-main/components/intent-toggle/index.tsx:
```diff
@@ -0,0 +1,53 @@
+	return (
+		<div className="intent-toggle">
+			<ToggleGroupControl
+				label={ translate( 'Plan type selector' ) }
```
I think this should be named something more relevant, as it's not the "plan type/term selector".

dzver replied:
It's an accessibility copy and was sent for translation so I'd rather not update it

chriskmnds commented on client/my-sites/plans-features-main/components/intent-toggle/index.tsx:
```diff
@@ -0,0 +1,53 @@
+				label={ translate( 'Plan type selector' ) }
+				hideLabelFromVision
+				value={ currentIntent || '' }
+				onChange={ handleToggle as ( value: string | number | undefined ) => void }
```
Is it possible to avoid the inline type casting? 🤔

chriskmnds commented on client/my-sites/plans-features-main/components/intent-toggle/style.scss:
```diff
@@ -0,0 +1,20 @@
+@import "@wordpress/base-styles/breakpoints";
+@import "@wordpress/base-styles/variables";
+
+.intent-toggle {
```
NIT: Depending on where the package lives, we might want to define a more specific name, ideally prefixed with its parent i.e. if living inside `plans-grid-next`, then `plans-grid-next__intent-toggle`, just to avoid any potential conflict.

chriskmnds commented on client/my-sites/plans-features-main/components/intent-toggle/style.scss:
```diff
@@ -0,0 +1,20 @@
+	width: 432px;
+	margin: 9px auto calc(#{$grid-unit-40} + 3px) auto;
+
+	.components-toggle-group-control {
```
It's generally not a great idea to be referencing the internal class of the imported component, which may change unexpectedly and lead to visual regressions here. I believe the component accepts a classname, so that can be leveraged instead to define a custom class and reference here.

chriskmnds commented on client/my-sites/plans-features-main/components/intent-toggle/style.scss:
```diff
@@ -0,0 +1,20 @@
+	.components-toggle-group-control {
+		width: 100%;
+		border-radius: 2px;
+		border: 1px solid #949494;
```
Is this by design? Perhaps we have a defined variable with the value, or at least something that closely matches? 🤔

dzver replied:
I couldn't find one that closely matches and it becomes ugly with the two grays that are close to it

chriskmnds commented on client/my-sites/plans-features-main/components/intent-toggle/style.scss:
```diff
@@ -0,0 +1,20 @@
+	// Responsive: full width on smaller screens
+	@media ( max-width: $break-small ) {
+		width: calc(100% - 32px); // 100% minus 16px on each side
+		margin: 0 auto 16px auto; // Center with bottom margin
```
I think same as
```suggestion
		margin: 0 auto 16px; // Center with bottom margin
```

chriskmnds commented on client/my-sites/plans-features-main/components/intent-toggle/style.scss:
```diff
@@ -0,0 +1,20 @@
+	}
+
+	// Responsive: full width on smaller screens
+	@media ( max-width: $break-small ) {
```
Is it possible to make this mobile-first? So to define the mobile styles as default and adjust them for wider views instead?

chriskmnds commented on client/my-sites/plans-features-main/components/plans-page-subheader.tsx:
```diff
@@ -10,7 +10,6 @@ import { SelectedFeatureData } from '../hooks/use-selected-feature';
 const Subheader = styled.p< { isUsingStepContainerV2?: boolean } >`
 	margin: -32px 0 40px 0;
 	color: var( --studio-gray-60 );
-	font-size: 1rem;
```
Is this intended or leftover code? (just as no explanation anywhere)

dzver replied:
Thanks, I removed it by accident. A previous attempt to implement the subheader involved lots of changes to this code. It also doesn't seem to do anything

chriskmnds commented on client/my-sites/plans-features-main/index.tsx:
```diff
@@ -467,10 +478,13 @@ const PlansFeaturesMain = ( {
 		[ gridPlansForFeaturesGridRaw, deemphasizeFreePlan ]
 	);
 
+	const isVisualSplitEnabled =
+		intent === 'plans-website-builder' || intent === 'plans-wordpress-hosting';
```
This same condition happens in a couple of places, I believe in 3 other files and repeated a few times in a single file. Is there a way to return `isVisualSplitEnabled` from the eligibility hook instead?

chriskmnds commented on config/development.json:
```diff
@@ -147,6 +147,7 @@
 		"plans/self-service-downgrade": false,
 		"plans/updated-storage-labels": true,
 		"plans/upgradeable-storage": true,
+		"playground": true,
```
Is this intended or just leftover code?

chriskmnds commented on packages/calypso-products/src/constants/features.ts:
```diff
@@ -351,6 +351,7 @@ export const FEATURE_POST_EDITS_HISTORY = 'post-edits-history';
 export const FEATURE_SECURITY_BRUTE_FORCE = 'security-brute-force';
 export const FEATURE_SMART_REDIRECTS = 'smart-redirects';
 export const FEATURE_ALWAYS_ONLINE = 'always-online';
+export const FEATURE_ALWAYS_ON_UPTIME = '99-999-uptime';
```
Should we instead name this similarly to the rest (with name matching the value)? e.g. like `always-online` above.

dzver replied:
feature renamed to FEATURE_99_999_UPTIME for consistency

chriskmnds commented on packages/plans-grid-next/src/hooks/data-store/use-plan-features-for-grid-plans.ts:
```diff
@@ -110,6 +110,32 @@ const usePlanFeaturesForGridPlans: UsePlanFeaturesForGridPlans = ( {
 							return true;
 						} );
 					}
+				} else if ( 'plans-wordpress-hosting' === intent ) {
+					// Use visual split features for WordPress hosting intent
+					if ( planConstantObj?.getVisualSplitBusinessFeatures ) {
+						wpcomFeatures = getPlanFeaturesObject(
+							allFeaturesList,
+							planConstantObj.getVisualSplitBusinessFeatures() ?? []
+						);
+					} else if ( planConstantObj?.getVisualSplitCommerceFeatures ) {
+						wpcomFeatures = getPlanFeaturesObject(
+							allFeaturesList,
+							planConstantObj.getVisualSplitCommerceFeatures() ?? []
+						);
+					} else {
+						// Fallback to default features if visual split features aren't available
+						wpcomFeatures = getPlanFeaturesObject(
+							allFeaturesList,
+							planConstantObj?.get2023PricingGridSignupWpcomFeatures?.( {
+								isSummerSpecial,
+							} ) ?? []
+						);
+					}
+
+					jetpackFeatures = getPlanFeaturesObject(
+						allFeaturesList,
+						planConstantObj.get2023PricingGridSignupJetpackFeatures?.() ?? []
+					);
```
Is this something that could be achieved by passing a parameter into `resolveFeatureGroupsForFeaturesGrid` instead? It looks like this is the current method/convention, seeing `isSummerSpecial` above used in a similar manner. 

It might avoid doing any conditioning here if it means simply passing the argument across suffices.

chriskmnds commented on packages/calypso-products/src/features-list.tsx:
```diff
@@ -1856,6 +1867,13 @@ const FEATURES_LIST: FeatureList = {
+
+	[ FEATURE_ALWAYS_ON_UPTIME ]: {
+		getSlug: () => FEATURE_ALWAYS_ON_UPTIME,
+		getTitle: () => i18n.translate( '99.999% uptime' ),
```
I think we should parse this with `@automattic/number-formatters` to ensure the `99.999%` is properly and uniformly translated. There are a few other cases of percentages being parsed with `number-formatters`.

chriskmnds commented on packages/plans-grid-next/src/components/features-grid/index.tsx:
```diff
@@ -178,7 +178,11 @@ const MobileView = ( {
+								<h2
+									className={ `plans-grid-next-features-grid__feature-group-title${
+										intent === 'plans-wordpress-hosting' ? ' is-wordpress-hosting' : ''
+									}` }
```
Can we use `clsx` for the class name? I think concatenating this way isn't great for maintenance/debugging, or consistency.

chriskmnds commented on packages/plans-grid-next/src/components/features-grid/plan-features-list.tsx:
```diff
@@ -90,7 +90,11 @@ const PlanFeaturesList = ( {
+								<h2
+									className={ `plans-grid-next-features-grid__feature-group-title${
+										intent === 'plans-wordpress-hosting' ? ' is-wordpress-hosting' : ''
+									}` }
```
Similar, if we can use `clsx`.

chriskmnds commented on packages/plans-grid-next/src/components/features-grid/plan-features-list.tsx:
```diff
@@ -136,7 +140,11 @@ const PlanFeaturesList = ( {
+							<h2
+								className={ `plans-grid-next-features-grid__feature-group-title${
+									intent === 'plans-wordpress-hosting' ? ' is-wordpress-hosting' : ''
+								}` }
```
Likewise, if we can use `clsx`.

chriskmnds commented on packages/plans-grid-next/src/components/features-grid/style.scss:
```diff
@@ -15,7 +15,7 @@
 
 	@include plans-grid-medium-large {
-		margin-top: 36px;
+		margin-top: 32px;
```
Is this uniformly expected i.e. it won't have unexpected results if the toggle is not rendered?

dzver replied:
Done on purpose. The 36px is inconsistent so rather than fixing with a separate PR, I just updated it to 32px everywhere.

chriskmnds commented on packages/plans-grid-next/src/components/features-grid/style.scss:
```diff
@@ -319,6 +319,11 @@
 	font-weight: 600;
 	color: var(--studio-gray-80);
+
+	&.is-wordpress-hosting {
+		padding-top: 12px;
+		font-size: 12px;
+	}
```
I am personally not a big fan that the modifier is leaking specific context from plans-intent into the plans grid styles. I think we tried to avoid that as much as possible in the package (at least in the process of refactoring, so not entirely). I think a better abstraction would be to focus on the actual behavior (or problem it solves) and build it out and name it from there. What problem is this addressing?

dzver replied:
<img width="387" height="110" alt="Screenshot 2025-09-02 at 15 49 41" src="https://github.com/user-attachments/assets/1ca9c0c2-90d8-45c7-aa59-5ca607120fe3" />

These group titles have different styling

chriskmnds commented on packages/plans-grid-next/src/hooks/data-store/use-grid-plans.tsx:
```diff
@@ -214,6 +214,12 @@ export const usePlanTypesWithIntent = ( {
 			// This plan intent is currently not utilized but will be soon
 			planTypes = [ TYPE_PREMIUM, TYPE_BUSINESS, TYPE_ECOMMERCE ];
 			break;
+		case 'plans-wordpress-hosting':
+			planTypes = [ TYPE_BUSINESS, TYPE_ECOMMERCE, TYPE_ENTERPRISE_GRID_WPCOM ];
+			break;
+		case 'plans-website-builder':
+			planTypes = [ TYPE_FREE, TYPE_PERSONAL, TYPE_PREMIUM, TYPE_BUSINESS ];
+			break;
```
Amazing to see plan-intent coming in handy again. 👍

dzver replied:
Thanks

dzver committed:
Sentence casing requested by Tess

dzver committed:
Remove one use of the variation that is not needed

Props Claudiu

dzver committed:
The Figma has no free link in the hosting tab

dzver committed:
Add a way to trigger the modal from outside of PlansFeaturesMain

chriskmnds commented on client/my-sites/plans-features-main/components/intent-toggle/index.tsx:
NIT: If this is a generic component, then we could instead bring it inside plans-grid-next package to allow future reuse (such as in LandPack or the v2 dashboard). 

I would also consider making it context-agnostic e.g. have it accept an array of values or index like `{ [plans-intent]: [label] }`. Then it becomes strictly presentational.

dzver replied:
Thanks, this can be done in case the experiment succeeds

chriskmnds commented on client/my-sites/plans-features-main/components/intent-toggle/index.tsx:
```diff
@@ -0,0 +1,53 @@
+	onIntentChange: ( intent: PlansIntent ) => void;
+}
+
+export default function IntentToggle( { currentIntent, onIntentChange }: IntentToggleProps ) {
```
Maybe `PlansIntentToggle`? Also updating the classname below.

dzver committed:
Unbreak the modal

chriskmnds reviewed (changes requested):
Heya. Pretty awesome work. I reviewed the code and left a few inline code comments, mostly suggestions. Also,  an edge-case regression below that I noticed while testing.

Just as a general suggestion, I think the PR tries to compile a little too much in one go. It makes reviewing/catching errors more difficult, and also may not be ideal for referencing back down the line. There could be some natural splits that could have been applied: (1) the eligibility logic, (2) the intent-toggle and intent definition, (3) the feature groups adjustment. and probably more fine-tuning on top.

Not unrelated, the PR description is also a little vague. A reference to the issue is also missing. Thinking of someone looking back to confirm some CSS change or logic, it would be hard. 

#### Regression:

The "view all plans" escape hatch is broken now while testing the admin page with a tailored grid e.g. `/plans/:site` with a newsletter site. It leads to a blank page / infinite loop in my testing. This sounds related to the changes in the intent resolution in plans-features-main.

<img width="500" alt="Screenshot 2025-09-02 at 12 22 32 PM" src="https://github.com/user-attachments/assets/4e68df91-9e04-452e-961a-dbbf0a71dc56" />

dzver committed:
Apply styles to the modal

dzver committed:
Remove rebase artifact

dzver committed:
use clsx per request by Christos

dzver committed:
Missed styles for the hosting intent

dzver committed:
Use clsx per request

dzver committed:
Remove styles from step container per Christos' request

dzver committed:
Rename class

dzver committed:
Fix regression spotted by Christos

The fix was created by Claudiu

dzver committed:
Rename feature per request

dzver committed:
Reduce dependebility on styles

dzver committed:
No inline casting

dzver committed:
Mobile first style

dzver committed:
Restore 1rem code

I accidentally removed it and it doesn't do anything so nothing changed
and I didn't notice

dzver committed:
Format the % per request

dzver committed:
Move file to lib/plans instead of lib/domains

dzver committed:
There's no longer a need to modify this file

dzver committed:
Remove 17px negative margin, no longer does anything useful

breaks the features list on mobile

dzver committed:
The title Storage is hidden in the design

dzver commented:
> Just as a general suggestion, I think the PR tries to compile a little too much in one go. It makes reviewing/catching errors more difficult, and also may not be ideal for referencing back down the line. There could be some natural splits that could have been applied: (1) the eligibility logic, (2) the intent-toggle and intent definition, (3) the feature groups adjustment. and probably more fine-tuning on top.

I'd like to address that as a pre-mortem because a change that large can cause problems.

The deadline was very tight. I started working with the intent of breaking it up in pieces but then the design and copy changes kept flowing until Friday last week with a target launch of Tuesday. I felt, right or not, that it would be unmanageable with multiple PRs. One merge conflict away of not making it on time. I also foolishly believed it won't get that big.

Probably saved 2-3 days of additional work by doing this and I hope any bugs that might slip in because of my large change would not be serious.
